dist/angular-form-messages-bootstrap.js

```diff
@@ -30,40 +29,99 @@ angular.module('angularFormMessagesBootstrap')
 angular.module('angularFormMessagesBootstrap')
 /**
  * This directive shows the 'has-error', 'has-warning', 'has-info' or 'has-success' when there is one message on the field.
- * Where there are multiple messages, no class is added.
+ * Where there are multiple messages and multiple messages are shown, no class is added.
  * When there is no message (the field is valid) and showSuccess is true, show the 'has-success' class.
  */
-  .directive('afFieldState', ["MESSAGE_TYPES", "MessageService", function (
-    MESSAGE_TYPES,
-    MessageService
+  .directive('afFieldState', ["$timeout", "AfMessageService", "MESSAGE_TYPES", function (
+    $timeout,
+    AfMessageService,
+    MESSAGE_TYPES
   ) {
     return {
-      require: ['^afSubmit', '^form'],
-      link: function ($scope, elem, attrs, ctrls) {
+      require: '^afSubmit',
+      controller: angular.noop,
+      link: function ($scope, elem, attrs, afSubmitCtrl) {
+        function typeToFieldStateClass(messageType) {
+          var mapping = {};
+          mapping[MESSAGE_TYPES[0]] = 'has-success';
+          mapping[MESSAGE_TYPES[1]] = 'has-info';
+          mapping[MESSAGE_TYPES[2]] = 'has-warning';
+          mapping[MESSAGE_TYPES[3]] = 'has-error';
+          return mapping[messageType];
+        }
+
+        function alertClassToType(alertClass) {
+          return {
+            'alert-success': MESSAGE_TYPES[0],
+            'alert-info': MESSAGE_TYPES[1],
+            'alert-warning': MESSAGE_TYPES[2],
+            'alert-danger': MESSAGE_TYPES[3]
+          }[alertClass];
+        }
+
+        function addClass(messageType) {
+          attrs.$addClass(typeToFieldStateClass(messageType));
+        }
+
+        function removeClass(messageType) {
+          attrs.$removeClass(typeToFieldStateClass(messageType));
+        }
+
+        function groupByMessageType(alerts) {
+          var groupedAlerts = {};
+          angular.forEach(alerts, function (alert) {
+            // Find the alert-* class, expects one found class
+            var type = alert.className.match(/(alert-\w+)/g)[0];
+            groupedAlerts[type] = alert;
+          });
+          return groupedAlerts;
+        }
+
         var
-          afSubmitCtrl = ctrls[0],
-          formCtrl = ctrls[1],
+          cachedMessages = [],
           messageId = attrs.afFieldState || attrs.afMessageId;
 
-        MessageService.validation(formCtrl.$name + '.' + messageId, function (messages, messageType) {
-          angular.forEach(MESSAGE_TYPES, function (type) {
-            attrs.$removeClass('has-' + type.toLowerCase());
-          });
+        $scope.$on('validation', function (event, validationMessageId, messages) {
+          // Make sure that the AfMessages validation callback runs first
+          $timeout(function () {
+            var alerts = elem[0].querySelectorAll('.alert');
+            var alertClasses = Object.keys(groupByMessageType(alerts));
```

Comment:
Mja IE8 werkt ook niet altijd met `querySelectorAll` en daarnaast is het toch een beetje op 1.3 gericht (dus niet meer IE8 compatibel). Liever dit dan dat ik Lodash er in moet hangen. Minder deps is beter.